[1-3단계 - 방탈출 예약 관리] 라젤(문선영) 미션 제출합니다.

RaZel713 opened the pull request:
# Mission 1) 방탈출 예약 관리 
안녕하세요~ 로키 리뷰어님! 😊
레벨 2 첫 번째 미션을 리뷰이로 함께 하게 된 라젤이라고 합니다!

레벨 1에서 다른 크루들을 통해 로키 리뷰어님을 알게 되면서 꼭 한번 리뷰를 받아 보고 싶었는데, 
이렇게 레벨 2 첫 미션에서 리뷰이로 만날 수 있게 되어 너무 기쁩니다!
부족하지만 잘 부탁드립니다! 🥰

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
페어와 여러 번 체크하면서 프로그래밍 하였기 때문에 잘 충족하였다고 생각합니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [x] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

RaZel713 committed:
init: 초기 연결 설정

RaZel713 committed:
docs: step1 README 작성

RaZel713 committed:
feat: 어드민 페이지 응답 구현

changuii committed:
feat: 웰컴 페이지 응답 구현

changuii committed:
docs: 2단계 요구사항 정의

changuii committed:
feat: admin/reservation 요청시 예약 페이지 응답 구현

RaZel713 committed:
feat: 예약 조회 기능 구현

changuii committed:
docs: 3단계 미션 요구사항 작성

changuii committed:
feat: 예약 생성 기능 구현

RaZel713 committed:
feat: 예약 삭제 기능 구현

changuii committed:
refactor: Admin 페이지는 AdminCotroller에 작성

changuii committed:
refactor: ReservationController 리팩터링

changuii committed:
refactor: 패키지 분리

changuii committed:
test: 테스트 패키지 분리

changuii committed:
fix: getter()가 없어 reservation이 json으로 잘못 변환되는 문제 해결

changuii committed:
refactor: final 키워드 추가

RaZel713 committed:
docs: 요구 사항만 작성하도록 변경

RaZel713 committed:
refactor: ReservationController 의 메서드 명을 더 자세히 작성

RaZel713 committed:
refactor: ReservationController의 메서드 순서 변경

- POST, GET, PUT, PATCH, DELETE 순으로 정렬

RaZel713 committed:
refactor: 예약 로직을 Reservations 객체가 관리

RaZel713 committed:
feat: 주어진 아이디에 해당하는 예약이 없다면 404로 응답

changuii committed:
test: 예약 api들을 분리해서 테스트 작성

- 생성, 조회, 삭제 기능들을 각각 분리해서 호출 테스트

changuii committed:
test: api 호출 테스트를 위한 클래스 이름을 명확하게 변경

changuii committed:
chore: 클래스 선언과 첫 필드 변수 사이 개행 처리

RaZel713 committed:
refactor: Reservation을 불변 객체로 변환

RaZel713 committed:
refactor: 네이밍 변경 및 불필요한 코드 제거

Rok93 commented on README.md:
````diff
@@ -0,0 +1,84 @@
+
+// Response
+HTTP/1.1 200
+```
````
각 API별로 요청 응답 예시, 응답 예시 등을 잘 기재해주셨네요. 👍💯

 IntelliJ IDE는 [Http Client](https://jojoldu.tistory.com/266)라는 기능을 지원한답니다. 🤗 
이런 기능으로 각 API에 대한 예시를 테스트하기 쉽게 만들 수도 있을 것 같아요. 😃

* [Http Client - JetBrain 소개](https://www.jetbrains.com/help/idea/http-client-in-product-code-editor.html#composing-http-requests)

RaZel713 replied:
좋은 기능을 알려주셔서 감사합니다! 
Http Client를 활용하여 `/src/api `에 관련 파일을 생성하여
API에 대한 예시를 테스트할 수 있도록 추가하였습니다!

Rok93 commented on src/test/java/roomescape/reservation/ReservationApiTest.java:
```diff
@@ -0,0 +1,93 @@
+        RESERVATION_BODY.put("name", "브라운");
+        RESERVATION_BODY.put("date", "2023-08-05");
+        RESERVATION_BODY.put("time", "15:40");
+    }
```
어짜피 정해진 상수고 이후로 변경되지 않는다면, 처음부터 상수로 바로 선언해도 되지 않을까요? 
`@BeforeAll`을 활용하여 초기화해주신 이유가 있을까요? 😃 


추가적으로 이런 경우에는 [맵 팩토리](https://devbksheen.tistory.com/entry/%EB%AA%A8%EB%8D%98-%EC%9E%90%EB%B0%94-%EC%BB%AC%EB%A0%89%EC%85%98-%ED%8C%A9%ED%86%A0%EB%A6%ACCollection-Factory-%EC%86%8C%EA%B0%9C-%EB%B0%8F-%EC%82%AC%EC%9A%A9%EB%B2%95#%EB%A7%B5%20%ED%8C%A9%ED%86%A0%EB%A6%AC(Map%C2%A0Factory)-1)를 활용해봐도 좋을 것 같아요. 😃 

```suggestion
    private static final Map<String, String> RESERVATION_BODY = Map.of(
            "name", "브라운",
            "date", "2023-08-05",
            "time", "15:40"
        );
```

RaZel713 replied:
해당 부분에 대하여 생각해보지 못한 것 같습니다! 알려주셔서 감사합니다!😊

Rok93 commented on src/main/java/roomescape/reservation/Reservations.java:
```diff
@@ -0,0 +1,36 @@
+
+public class Reservations {
+
+    private final AtomicLong nextId = new AtomicLong(1);
```
추가적으로 그냥 Long을 쓸 수도 있을 것 같은데, AtomicLong을 쓴 이유가 있을까요? 😃 
(혹시 라젤의 선택이 아닐지라도..! 이참에 알아가보면 좋을 것 같아요)

Rok93 replied:
AtomicLong과 같은 장점을 `List<Reservation>`의 경우에도 비슷하게 개선해보면 어떨까요?

RaZel713 replied:
현재 상황(단일 스레드 환경)에서는 동시성 문제가 발생하지 않지만, **실제로 서비스되는 상황(멀티 스레드 환경)에서는 동시성 문제가 발생하기 때문에 AtomicLong을 사용하였습니다!**

**`AtomicLong`** 은 Long 자료형을 가지고 있는 Wrapping 클래스입니다.
Thread-safe로 구현되어 멀티스레드에서 현재 데이터를 사용하고 있는 해당 스레드를 제외하고 나머지 스레드들은 데이터에 접근 할 수 없도록 하여 안정성과 신뢰성을 보장할 수 있습니다.

또한 같은 역할을 하는 **`synchronized`** 키워드보다 적은 비용으로 동시성을 보장할 수 있습니다.

<details>
<summary>추가 학습) synchronized보다 왜 적은 비용일까?</summary>

`synchronized` 는 JVM 수준에서의 모니터 락을 사용합니다. 즉, 스레드가 해당 블록을 들어가기 위해서는 락을 획득하고 다른 스레드는 대기해야 함을 의미합니다.

`AtomicLong`은 CAS (Compare-And-Swap) 기술을 사용합니다. 이것은 락 없이 원자적 연산을 수행할 수 있게 해주는 저수준 연산입니다. 즉 스레드가 대기하지 않고 바로 실패하게 되고 다시 시도할 수 있게 됩니다. (스핀 락 방식) 

따라서 멀티스레드 환경에서 단순한 숫자 증가 용도라면 AtomicLong이 synchronized보다 일반적으로 더 효율적입니다.
</details>

RaZel713 replied:
동시성 문제를 고려하여 List<Reservation>을 수정하는데 사용되는 메서드에 `synchronized` 키워드를 사용하여 멀티 스레드 환경에서 예상치 못한 문제가 발생하지 않도록 하였습니다!

<details>
<summary>synchronized와 AtomicLong 모두 사용한 이유</summary>

nextId는 List<Reservation>을 수정하는데 사용되는 메서드 내부에서 사용되는 값이므로, 
synchronized가 이미 멀티스레드 접근을 직렬화 하고 있어 AtomicLong를 사용해야 하는 이유는 없습니다. 
하지만, 확장성을 고려하여 둘 모두 사용하였습니다.
</details>

동시성 문제를 고려하여 AtomicLong를 사용한 것이 아니라, 학습 테스트에서 나와 있어 그대로 사용하게 된 것 같습니다! 
리뷰어님께서 짚어 주셔서 학습할 수 있는 좋은 계기가 된 것 같습니다! 감사합니다 😊

Rok93 replied:
> 현재 상황(단일 스레드 환경)에서는 동시성 문제가 발생하지 않지만, 실제로 서비스되는 상황(멀티 스레드 환경)에서는 동시성 문제가 발생하기 때문에 AtomicLong을 사용하였습니다!

[Spring Web MVC 환경의 애플리케이션은 이미 멀티스레드 환경에서 동작](https://trillium.tistory.com/69)하기 때문에 동시성 상황은 지금의 상황에서도 충분히 발생할 수 있다고 생각해요. 😃

지금 단계에서 내부 동작원리 등을 자세하게 공부하실 필요는 없지만 가볍게 특징들을 찾아보시는 것도 도움이 되실 것 같습니다. 
지금은 스프링과의 친밀도를 쌓는 과정이라고 생각해주시고 힘 빼고 하나씩 알아가보면 좋을 것 같습니다. 😉

Rok93 replied:
AtomicLong 에 대해서도 잘 답변해주신 것 같아요. 👍👍 
`synchronized` 키워드에 대한 추가 학습도 훌륭합니다! 💯🤩 

이번 기회에 공부하고 남겨주신 내용들은 짧게라도 블로그나 WIL 등에 잘 정리해두시면 나중에 다시 보시게될 때 도움이 많이 되실거에요. 😉 

추가적으로 여유가 되신다면 [해당 코멘트](https://github.com/woowacourse/spring-roomescape-admin/pull/217/files#r2049788977)도 개선해보면 좋을 것 같아요. 😃

Rok93 commented on src/main/java/roomescape/reservation/Reservations.java:
```diff
@@ -0,0 +1,36 @@
+        return reservations.stream()
+                .filter(reservation -> Objects.equals(reservation.id(), id))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("[ERROR]"));
```
예외 클래스가 적절한지 고민해보시면 좋을 것 같아요. 🤔
표준 예외 중에서 조금 더 적절한 예외는 없을까요 ? 

추가적으로 예외 메시지도 조금 더 의미있게 변경해보면 좋을 것 같습니다. 😉

RaZel713 replied:
id에 해당하는 예약이 존재하지 않는 경우이므로, 
아래의 3가지 예외 중 **NoSuchElementException**가 적합하다고 생각해 예외 메시지를 의미있게 변경하여 
리팩터링 하였습니다.

1. **NoSuchElementException**
컬렉션이나 Optional에서 원하는 요소가 존재하지 않을 때 사용하는 예외입니다.
현재처럼 findFirst().orElseThrow()와 함께 사용하면 적합하다고 생각합니다.

2. EntityNotFoundException
JPA 환경에서 엔티티가 존재하지 않을 때 사용하는 예외입니다.
현재 코드는 JPA 기반이 아니고, 메모리 컬렉션 기반에서 예약 데이터를 찾는 것이므로 어색하다고 생각합니다.

3. IllegalArgumentException
메서드에 전달된 인자가 부적절할 때 사용하는 예외입니다.
현재는 “id 값 자체가 잘못된 것”보다는 “해당 id의 예약이 존재하지 않는 것”에 가까우므로, 의미상 어색하다고 생각합니다.

Rok93 commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,20 @@
+    public Reservation {
+        Objects.requireNonNull(name);
+        Objects.requireNonNull(date);
+        Objects.requireNonNull(time);
```
name, date, time 값으로 null이 들어오면 어떻게될까요? 🤔
해당 예외가 의도한 예외가 맞으실까요? 👀

<img width="480" alt="image" src="https://github.com/user-attachments/assets/300eb986-3191-4aed-b896-95f4625d58f2" />

또한 예외가 발생했을 때, 왜 예외 메시지를 명확하게 던져주는게 추후 에러를 파악하기가 수월할 것 같아요. 😃
예외 메시지를 적절히 지정해주시면 어떨까 싶어요.

RaZel713 replied:
두번째 미션인 방탈출 사용자 예약에서 예외 처리 관련 사항을 다룬다고 생각하여, 
예약을 생성하는데 있어 가장 기본적인 null 체크만 간단하게 하고자 하였습니다.

하지만 간단히 처리한다고 해도 적절한 예외 메시지와 정확한 에러를 던지는 것이 중요하다는 점을 신경 쓰지 못했던 것 같습니다.

해당 부분에 대하여 반성하며, 
사용자가 적절한 입력을 넣지 않는 경우이므로, **IllegalArgumentException**을 사용하고,
어떤 값(name, date, time)에 null이 들어왔는지 명확하게 알 수 있도록 예외 메시지를 작성하여 예외처리 하였습니다.

```java
public Reservation {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("[ERROR] 이름이 비어 있을 수 없습니다.");
        }
        if (date == null) {
            throw new IllegalArgumentException("[ERROR] 날짜가 비어 있을 수 없습니다.");
        }
        if (time == null) {
            throw new IllegalArgumentException("[ERROR] 시간이 비어 있을 수 없습니다.");
        }
}
```

Rok93 commented on src/main/resources/static/index.html:
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
```
<img width="719" alt="image" src="https://github.com/user-attachments/assets/57eab9ba-1201-417e-82e7-f83e351a235b" />


이 index.html 파일이 지운 뒤 실행해보신 뒤, `template engine`
한번 해당 파일을 삭제한 뒤, 실행해보시면 좋을 것 같아요. 

[template engine](https://hi-hahahoho.tistory.com/99)에 대해서 가볍게 읽어보셔도 좋을 것 같네요. 😃

RaZel713 replied:
```md
// 추가 요구사항
- `localhost:8080` 요청 시 welcome 페이지가 응답할 수 있도록 구현한다.
    - welcome 페이지는 `static/index.html` 파일을 이용한다.

// 기존 요구사항
- `localhost:8080/admin`요청 시 어드민 메인 페이지가 응답할 수 있도록 구현한다.
    - 어드민 메인 페이지는 `templates/admin/index.html` 파일을 이용한다.
```

요구사항에 **localhost:8080/admin** 요청 시 어드민 메인 페이지를 응답해야 한다는 내용은 있었지만,
**localhost:8080** 요청 시 어떤 페이지가 출력되어야 하는지는 명시되어 있지 않았습니다.

기본 페이지는 항상 존재해야 한다고 판단하여, 기존 html에 인사 멘트를 추가해 기본 페이지(static/index.html)를 구현했습니다.
네이버도 도메인 주소만 입력해도 메인 페이지가 뜨는 것처럼 기본 페이지가 있어야 한다고 생각했습니다!

혹시 다른 점을 이야기 해 주시려는 의도이셨을까요? 🤔

Rok93 replied:
> localhost:8080 요청 시 어떤 페이지가 출력되어야 하는지는 명시되어 있지 않았습니다.

아... ! 저도  미션 요구사항에 welcome page에 대한 요구사항은 없어서 불필요한 구현이라 생각했었는데, 힌트에는 또 이런 내용이 있었군요? 🤔 welcome page 등록도 괜찮은 것 같아요. 👍

다른점을 이야기하려는 것은 아니였어요. 😃

Rok93 commented on src/main/java/roomescape/admin/AdminController.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+@RequestMapping("/admin")
```
endPoint의 prefix가 /admin인 경우에는 응답이 String이고 /reservations인 경우에는 ResponseEntity를 반환하고있는데요. 반환 타입이 다른 이유는 무엇인가요? 😃

또 `@ResponseBody`라는 애너테이션에 대해서도 공부해보면 좋을 것 같아요. 😃

RaZel713 replied:
반환 타입이 다른 이유는 담당하는 역할이 다르기 때문입니다! 

endPoint의 prefix가 /admin인 경우는 템플릿 뷰를 반환하고
endPoint의 prefix가 /reservations인 경우는 데이터를 JSON으로 반환합니다.

템플릿 뷰를 반환하는 메서드들은 AdminController에, 
JSON 데이터를 반환하는 메서드들은 ReservationController에 배정하여 
컨트롤러마다 명확한 책임과 응답 형식을 유지하도록 작성하였습니다!😁

RaZel713 replied:
**`@ResponseBody` 애너테이션이란?**
Spring MVC에서 컨트롤러 메서드의 반환 값을 HTTP Response Body에 직접 쓰도록 지정하는 역할을 합니다. 
즉, **뷰 리졸버를 거치지 않고 데이터를 직접 반환**하게 만듭니다.

반환 값은 보통 JSON 형식으로 직렬화 되고, **HTML 뷰가 아닌 데이터만 보내고 싶을 때** 사용합니다.

➡️`@Controller` 클래스나 메서드에 붙이면, 메서드의 반환 값을 HTTP 응답의 본문으로 직렬화 해서 클라이언트에 전달할 수 있습니다.

<details>
<summary>뷰 리졸버(View Resolver)란?</summary>
Spring MVC에서 컨트롤러가 반환한 뷰 이름(String)을 가지고 실제 뷰 파일(HTML, JSP 등) 을 찾아주는 중간 처리기
</details>

<details>
<summary>직렬화란?</summary>
객체를 바이트 형태나 텍스트로 변환해서 저장하거나 전송 가능하게 만드는 과정

객체를 JSON (텍스트)으로 변환하는 것이 직렬화!
</details>

Rok93 commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<Reservation>> readAllReservation() {
```
추가적으로 `@RestController` 키워드에 대해 학습해보시면 좋을 것 같습니다. 😃

추가적으로 `@RestController`, `@Controller`, `@ResponseBody` 애너테이션의 코드를 살펴보면 좋을 것 같아요. 😉

RaZel713 replied:
**`@RestController`** 란?
`@Controller` + `@ResponseBody`를 합친 애너테이션입니다.
즉, 클래스 내 모든 메서드가 자동으로 JSON 형태로 반환되도록 설정해 줍니다.
주로 RESTful API를 만들 때 사용합니다.
```java
@RestController
@RequestMapping("/api")
public class ApiController {
    @GetMapping("/hello")
    public String hello() {
        return "Hello!";
    }
}
// -> 응답: "Hello!"
```

**`@Controller`** 란?
Spring MVC의 기본적인 컨트롤러 역할을 하는 애너테이션입니다.
뷰 이름(String) 을 반환하고, 뷰 리졸버(ViewResolver)가 뷰 이름에 맞는 HTML 등을 렌더링합니다.

```java
@Controller
public class PageController {
    @GetMapping("/hello")
    public String hello() {
        return "hello"; // hello.html 렌더링
    }
}
```

Rok93 commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @PostMapping
+    public ResponseEntity<Reservation> createReservation(
+            @RequestBody final Reservation reservation
```
`@RequestBody` 키워드에 대해서도 학습해보고 가볍게 코멘트 달아주시면 좋을 것 같아요. 

심화 학습거리를 같이 제공해드리자면...!
`@RequestParam`, `@ModelAttribute` 등의 애너테이션도 같이 찾아보시면 도움이되실 것 같습니다.

RaZel713 replied:
**`@RequestBody`란?**
HTTP 요청 본문에 담긴 JSON 데이터를 Java 객체로 변환해줍니다.
주로 POST,PUT 요청에서 사용되며, Jackson 같은 메시지 컨버터를 통해 자동으로 역직렬화 됩니다.

<details>
<summary>역직렬화란?</summary>
JSON 등과 같은 텍스트 데이터를 Java 객체로 변환하는 과정

즉, 클라이언트가 보낸 JSON 데이터를 서버에서 사용할 수 있도록 객체로 바꿔주는 작업입니다.
</details>

**`@RequestParam`란?**
쿼리 파라미터, 폼 데이터 등에서 값을 추출합니다. 
`?name=라젤`과 같이 URL에서 포함된 값을 매핑합니다.
```java
@GetMapping("/search")
public String search(@RequestParam String name) {
    ...
}
```

**`@ModelAttribute`란?**
폼 데이터를 객체로 바인딩할 때 사용됩니다.
`@RequestParam`와 달리 객체 단위로 묶어서 처리 가능할 수 있습니다.
내부적으로 set___() 메서드를 이용해 바인딩합니다.
```java
@PostMapping("/form")
public String submit(@ModelAttribute User user) {
    ...
}
```

`@RequestParam`과 `@ModelAttribute`의 차이점
➡️`@RequestParam`는 직접 파라미터 이름을 키로 값을 추출하지만, `@ModelAttribute`는 객체를 생성하고 setter로 값을 주입합니다.

Rok93 replied:
전체적으로 질문 드린부분에 대해서 답변 잘 남겨주셨네요. 👍💯
생각보다 낯선 개념들이 많이 포함되어있었을 것 같은데도, 잘 정리해주셨습니다. 

직렬화, 역직렬화, 뷰 리졸버 등등 중간중간 생소한 키워드에 대한 추가로 학습하신 내용 남겨주신 것들도 너무 좋습니다. 👍💯

Rok93 reviewed (changes requested):
라젤 안녕하세요. 레벨2 첫번째 미션을 함께하게된 로키입니다. 반가워요! 🤗
웹 애플리케이션 개발 경험이 없으심에도 1단계 훌륭하게 잘 구현해주셨네요. 👍👍  

크게 구현에 대한 피드백은 드릴 부분이 없었고, 몇몇 질문거리를 남겨두었습니다. 😃
관련해서 한번 고민해보고 학습해보시면서 가능한 코멘트로 답변도 남겨주시면 좋을 것 같아요. 
답변과 함께 리뷰요청주시면 1단계는 머지하도록하겠습니다. 😉

<img width="504" alt="image" src="https://github.com/user-attachments/assets/c23a6658-8afc-4edc-a578-00c57f6950a8" />

RaZel713 committed:
test: HTTP 요청 정리 파일 생성

- welcome, 어드민 메인, 예약 관리 페이지
- 예약 생성, 삭제, 조회

RaZel713 committed:
refactor: 예약 정보 입력 값이 null일 경우 구체적인 예외 메시지를 던지도록 수정

- 예약 생성 시 name, date, time의 null 여부를 검증하고 IllegalArgumentException으로 명확한 예외 메시지 반환하도록 변경한다.

RaZel713 committed:
feat: synchronized 키워드를 사용하여 Reservations 클래스에 동시성 처리 적용

RaZel713 committed:
refactor: findBy 메서드에서 던지는 예외 클래스 변경

- 적절한 예외 메시지를 작성한다.

RaZel713 committed:
refactor: 테스트 데이터 상수 처리

RaZel713 reviewed (commented):
안녕하세요~ 로키! 😊
리뷰를 너무 잘해주셔서 학습하는데 도움이 많이 되었어요!
답글을 너무 길지 않게 핵심만 작성해보려고 노력해 보았는데 **(comment도 code다!)**
조금 더 자세하게 적어야 하는지, 조금 더 요약하면 좋을 지 알려주시면 좋을 것 같아요~!

Rok93 commented on src/main/java/roomescape/reservation/Reservations.java:
```diff
@@ -0,0 +1,37 @@
+        return reservations.stream()
+                .filter(reservation -> Objects.equals(reservation.id(), id))
+                .findFirst()
+                .orElseThrow(() -> new NoSuchElementException("[ERROR] 해당 ID의 예약을 찾을 수 없습니다. id:" + id));
```
[잘 고민해서 선택](https://github.com/woowacourse/spring-roomescape-admin/pull/217#discussion_r2051633085)해주신 것 같아요. 👍 
단순히 이게 적절해보여요! 같은 방식이 아닌 여러 예외 클래스를 비교해보면서 적절한 예외 클래스를 도출해주신 의사결정의 흐름이 보기좋네요. 💯😃


그나저나..  뉴비라고하셨는데... JPA...? 🙄
뉴비에게서 익숙한 향기가....? 🤔

<img width="741" alt="image" src="https://github.com/user-attachments/assets/8448f11a-5b0a-4cfd-9a21-732b5e693ba5" />

Rok93 reviewed (approved):
라젤 질문에 대한 답변들 꼼꼼하게 잘 학습해서 답변남겨주셨네요. 👍💯
1단계는 충분히 잘 진행해주신 것 같아서 이만 머지하도록하겠습니다. 

다음 단계에서 뵐게요. 😉
